### PR TITLE
updates to events, fixes to avoid undefined behavior

### PR DESCRIPTION
**Pull request by slwu89**
This PR addresses #138 as well as making other changes to the events.

1. class methods for `Event` and `TargetedEvent` are defined outside of the class in inst/include/Event.h
2. add checks for input in R/targeted_event.R to avoid undefined behavior; specifically check for positive, finite input for arguments that will be coerced to `vector<size_t>` by Rcpp (to avoid CRAN UBSCAN errors).
3. add new method overload to `TargetedEvent::schedule` taking a bitset and vector of delays, so that a cohort with unique delay values can be scheduled without converting it to vector first.
4. add new function `round_double` to inst/include/Event.h that checks an input double is positive and finite before rounding and casting to `size_t`; now `round_delay` calls this internally. This helps users because when strange delays (`Inf`, `NaN`, etc) get passed, a more informative error message is printed.
5. in `event.cpp` when inserting into a bitset from a vector (where there is the possibility of the vector having bad values outside the range of the bitset), use `insert_safe` instead of `insert` to produce more informative error messages (this may be related to #35). I think this is an improvement, especially since it means we don't have to do a slower check at the R level, and insertion into the bitset means we have to iterate through the index vector anyway, so it should be faster and safer.
6. split `test-event.R` into 2 files, `test-event.R` and `test-targetedevent.R`; add many tests to catch corner cases that were not getting tested for targeted events.
7. For all types that can accept the `index` argument as a vector (`TargetedEvent`, `CategoricalVariable`, `IntegerVariable`, `DoubleVariable`), consistently check for positive, finite input before passing to C++ to avoid undefined behavior if values outside of the representable range of `size_t` type get passed (same as 2 but for all objects that might suffer from this problem).

I also bumped the version number and updated NEWS.md. Not sure what you think about updating the version @giovannic, none of the items here are especially urgent, but they were lingering things I wanted to get to. I checked this branch for undefined behavior and/or memory leaks using the docker container in Rich's blog post and confirmed none of either before submitting this PR.

### CODE REVIEW

File: inst/include/Event.h [hunk 1]
```diff
@@ -17,149 +17,269 @@
...
+}
 
-    virtual void clear_schedule(const individual_index_t& target) {
-        auto not_target = ~target;
-        for (auto& entry : targeted_schedule) {
-           entry.second &= not_target;
-        }
+//' @title clear scheduled events for `target` individuals
+inline void TargetedEvent::clear_schedule(const individual_index_t& target) {
+    auto not_target = ~target;
```
Comment (giovannic) [inline inst/include/Event.h:268]: I really regret making `~` the inverse operator for bitsets. Since it makes me think of destructors. It's out of scope for this though, so I've made an issue #141

File: tests/testthat/test-events.R [hunk 2]
```diff
@@ -15,66 +15,28 @@ test_that("first event is triggered at t=1", {
...
-  expect_targeted_listener(listener, 1, t = 3, target = c(2, 4))
 })
 
-test_that("events can be scheduled for for a Real time", {
-  event <- TargetedEvent$new(10)
+test_that("events can be scheduled for the future", {
+  event <- Event$new()
   listener <- mockery::mock()
   event$add_listener(listener)
-  event$schedule(c(2, 4), 1.9)
```
Comment (giovannic) [inline tests/testthat/test-events.R:77]: It would still be nice to check/show that the discretisation of time is intentional and works as expected

**Review by giovannic — CHANGES_REQUESTED** [review]
Thank you, thank you, thank you

File: R/double_variable.R [hunk 3]
```diff
@@ -20,13 +20,16 @@ DoubleVariable <- R6Class(
     #' or integer vector, return values of those individuals.
     get_values = function(index = NULL) {
       if (is.null(index)) {
-        return(double_variable_get_values(self$.variable))
-      }
-      if (is.numeric(index)) {
-        stopifnot(all(index > 0))
-        return(double_variable_get_values_at_index_vector(self$.variable, index))
+        double_variable_get_values(self$.variable)
```
[the remark this answers is not in the document]
Comment (slwu89) [reply R/double_variable.R]: addressed in https://github.com/mrc-ide/individual/pull/139/commits/396c7ec420801f5b57374d9a29acf9ca11dbf175

File: R/double_variable.R [hunk 4]
```diff
@@ -20,13 +20,16 @@ DoubleVariable <- R6Class(
...
+      } else {
+        if (inherits(index, 'Bitset')) {
+          double_variable_get_values_at_index(self$.variable, index$.bitset)
```
[the remark this answers is not in the document]
Comment (slwu89) [reply R/double_variable.R]: addressed in https://github.com/mrc-ide/individual/pull/139/commits/396c7ec420801f5b57374d9a29acf9ca11dbf175

File: R/double_variable.R [hunk 5]
```diff
@@ -20,13 +20,16 @@ DoubleVariable <- R6Class(
...
+        } else {
+          stopifnot(all(is.finite(index)))
+          stopifnot(all(index > 0))
+          double_variable_get_values_at_index_vector(self$.variable, index)
```
[the remark this answers is not in the document]
Comment (slwu89) [reply R/double_variable.R]: addressed in https://github.com/mrc-ide/individual/pull/139/commits/396c7ec420801f5b57374d9a29acf9ca11dbf175

File: R/integer_variable.R [hunk 6]
```diff
@@ -24,13 +24,17 @@ IntegerVariable <- R6Class(
     #' or integer vector, return values of those individuals.
     get_values = function(index = NULL) {
       if (is.null(index)) {
-        return(integer_variable_get_values(self$.variable))
-      }
-      if (is.numeric(index)) {
-        stopifnot(all(index > 0))
-        return(integer_variable_get_values_at_index_vector(self$.variable, index))
+        integer_variable_get_values(self$.variable)
```
[the remark this answers is not in the document]
Comment (slwu89) [reply R/integer_variable.R]: addressed in https://github.com/mrc-ide/individual/pull/139/commits/396c7ec420801f5b57374d9a29acf9ca11dbf175

File: R/integer_variable.R [hunk 7]
```diff
@@ -24,13 +24,17 @@ IntegerVariable <- R6Class(
...
+      } else{
+        if (inherits(index, 'Bitset')){
+          integer_variable_get_values_at_index(self$.variable, index$.bitset)
```
[the remark this answers is not in the document]
Comment (slwu89) [reply R/integer_variable.R]: addressed in https://github.com/mrc-ide/individual/pull/139/commits/396c7ec420801f5b57374d9a29acf9ca11dbf175

File: R/integer_variable.R [hunk 8]
```diff
@@ -24,13 +24,17 @@ IntegerVariable <- R6Class(
...
+        } else {
+          stopifnot(all(index > 0))
+          stopifnot(all(is.finite(index)))
+          integer_variable_get_values_at_index_vector(self$.variable, index)
```
[the remark this answers is not in the document]
Comment (slwu89) [reply R/integer_variable.R]: addressed in https://github.com/mrc-ide/individual/pull/139/commits/396c7ec420801f5b57374d9a29acf9ca11dbf175

File: inst/include/Event.h [hunk 9]
```diff
@@ -17,149 +17,269 @@
...
+    // for that delay if the delay matches the unique value
+    for (auto v : delay_values) {
+        auto target = individual_index_t(size);
+        auto bitset_it = target_bitset.cbegin();
+        for (auto i = 0u; i < rounded.size(); ++i) {
+            if (rounded[i] == v) {
+                target.insert(*bitset_it);
             }
-            schedule(target, v);
+            ++bitset_it;
```
[the remark this answers is not in the document]
Comment (slwu89) [reply inst/include/Event.h:203]: thanks!

File: tests/testthat/test-events.R (hunk 2 above)
Comment (slwu89) [reply tests/testthat/test-events.R:77]: addressed in https://github.com/mrc-ide/individual/pull/139/commits/396c7ec420801f5b57374d9a29acf9ca11dbf175

**Review by giovannic — APPROVED** [review]
👍
